flux-jobs: support new "inactive_reason" output field and "endreason" format

chu11 opened the pull request:
Per discussion in #4946.  This adds a new field `{inactive_reason}` to get why a job is inactive and a new named format in `flux-jobs` to help the user get the reason why their job is inactive.

After some trial and error, came up with the following defined format for `flux-jobs`, which I call "endreason".  Open to suggestions for an alternate name.  Its the best I could come up with :-)

```
+                "{id.f58:>12} ?:{queue:<8.8} {username:<8.8} {name:<10.10+} "
+                "{status_abbrev:>2.2} {t_inactive!D:>19h} {inactive_reason}"
```

example

```
       JOBID USER     NAME       ST          T_INACTIVE INACTIVE-REASON
    ƒCWZ9cgX chu11    sleep       R                   - 
    ƒBwTFejy chu11    sleep       R                   - 
    ƒJw1KBu1 chu11    nosuchcom+  F 2023-03-31T12:45:43 command not found
    ƒGf7P5mh chu11    true       CD 2023-03-31T12:45:38 Exit 0
    ƒF7xGGpw chu11    false       F 2023-03-31T12:45:35 Exit 1
    ƒ3hsP1TD chu11    sleep      CA 2023-03-31T12:45:25 Canceled: foobar
    ƒ4VrUjRZ chu11    sleep      CA 2023-03-31T12:45:15 Canceled
```

the `ST` field is still there just so users can see those are running jobs and thus don't get just an empty `INACTIVE-REASON` without any context.  I initially wanted to include `T_SUBMIT` but I couldn't really fit it in and make the format pretty enough for my tastes :-)  Also, I didn't like two hyphens next to each other for `T_INACTIVE` and `INACTIVE-REASON`, so there's just one  ... eh??  :-)

One debate I had on the output from an exec error, I elect to handle it as a special case, outputting the special bash exit code message of "command not found" (exit 126) vs the actual output that might come out from the exception, "task 0 (host foobar): start failed: nosuchcommand: No such file or directory".  Could tweak that.

This PR is built on top of #5031

chu11 commented:
hmmmm, my attempt to be clever doesn't like the `vermin` python version check.

```
                # Unfortunately strsignal() is only available in                                                                            
                # Python 3.8 and up.  flux-core's minimum is 3.6.                                                                           
                # If strsignal() is not available, we hand check for                                                                        
                # several common signals before outputting something                                                                        
                # generic.                                                                                                                  
                                                                                                                                            
                if "strsignal" in dir(signal):                                                                                              
                    try:                                                                                                                    
                        sigdesc = signal.strsignal(signum)                                                                                  
                    except ValueError:                                                                                                      
                        sigdesc = f"Signaled {signum}"                                                                                      
                else:                                                                                                                       
                    if signum == signal.SIGINT:                                                                                             
                        sigdesc = "Interrupt"                                                                                               
                    elif signum == signal.SIGKILL:                                                                                          
                        sigdesc = "Killed"                                                                                                  
                    elif signum == signal.SIGUSR1:                                                                                          
                        sigdesc = "User defined signal 1"                                                                                   
                    elif signum == signal.SIGSEGV:                                                                                          
                        sigdesc = "Segmentation Fault"                                                                                      
                    elif signum == signal.SIGUSR2:                                                                                          
                        sigdesc = "User defined signal 2"                                                                                   
                    elif signum == signal.SIGTERM:                                                                                          
                        sigdesc = "Terminated"                                                                                              
                    else:                                                                                                                   
                        sigdesc = f"Signaled {signum}"                                                                                      
                return sigdesc                            
```

dunno if there is a workaround for this?  Couldn't figure one out via the project's website.  Guess we could just go with the fallthrough one and just add a TODO to come back to this when Python 3.8 is the minimum required version.

grondo commented:
> Guess we could just go with the fallthrough one and just add a TODO to come back to this when Python 3.8 is the minimum required version.

That seems most reasonable. BTW, from what I've gathered, the more Pythonic way to do this (for future reference) is with a `try: except` block, rather than `if strsignal in dir(signal)`.  e.g.:

```python
import signal
  
def strsignal(signum):
    if signum == signal.SIGINT:
        return "Interrupt"
    elif signum == signal.SIGKILL:
        return "Killed"
    elif signum == signal.SIGUSR1:
        return "User defined signal 1"
    elif signum == signal.SIGSEGV:
        return "Segmentation Fault"
    elif signum == signal.SIGUSR2:
        return "User defined signal 2"
    elif signum == signal.SIGTERM:
        return "Terminated"
    else:
        return f"Signal {signum}"

try:
    result = signal.strsignal(15)
except AttributeError:
    result = strsignal(15)

print(result)
```

Which is also better since it encapsulates the compatibility code into a function.

In fact, if we had many of these functions we could place them all into a module `flux.compat36`, and then do something like:
```python
try:
    from signal import strsignal
except ImportError:
    from flux.compat36 import strsignal
```

Then once we no longer need to support python3.6, the whole module could be removed.

grondo reviewed (commented):
Sorry, just got around to reviewing this one.

Main suggested change is to factor a long if block into its own function, which can then be separately tested.

chu11 commented on src/cmd/flux-jobs.py:
```diff
@@ -60,6 +60,13 @@ class FluxJobsConfig(UtilConfig):
+            "description": "Show why each job ended",
+            "format": (
+                "{id.f58:>12} ?:{queue:<8.8} {username:<8.8} {name:<10.10+} "
+                "{status_abbrev:>2.2} {t_inactive!D:>19h} {inactive_reason}"
```
Hmmm, I think the reason I went with `!D` may have been because we (I don't think) support the `h` hyphen output with `!d`.  So I sort of dislike this output for the running jobs.

```
       JOBID USER     NAME       ST   T_INACTIVE INACTIVE-REASON
    f7qUkw7d achu     sleep       R  Dec31 16:00 
    f7qRnxYx achu     sleep       R  Dec31 16:00 
    f7qRnxYw achu     sleep       R  Dec31 16:00 
    f7qQJyGb achu     sleep       R  Dec31 16:00 
    f7gRwKMZ achu     sleep       F  May10 15:10 User defined signal 1
    f7XHBkU7 achu     sleep       F  May10 15:10 Terminated
    f7GqYkFH achu     sleep      CA  May10 15:10 Canceled: foobar
    f72GVocj achu     sleep      CA  May10 15:10 Canceled
    f6tkdSmM achu     nosuchcom+  F  May10 15:10 command not found
    f6mBo7MH achu     false       F  May10 15:10 Exit 1
    f6dU4rF9 achu     true       CD  May10 15:10 Exit 0
```

Maybe we can tweak/handle `h` in `UtilDatetime`.

But along the way, I noticed this:

```
        if conv == "d":                                                                                                                     
            # convert from float seconds since epoch to a datetime.                                                                         
            # User can than use datetime specific format fields, e.g.                                                                       
            # {t_inactive!d:%H:%M:%S}.                                                                                                      
            try:                                                                                                                            
                value = UtilDatetime.fromtimestamp(value)                                                                                   
            except TypeError:                                                                                                               
                if orig_value == "":                                                                                                        
                    value = UtilDatetime.fromtimestamp(0.0)                                                                                 
                else:                                                                                                                       
                    raise                                                                                                                   
        elif conv == "D":                                                                                                                   
            # As above, but convert to ISO 8601 date time string.                                                                           
            try:                                                                                                                            
                value = datetime.fromtimestamp(value).strftime("%FT%T")                                                                     
            except TypeError:                                                                                                               
                if orig_value == "":                                                                                                        
                    value = ""                                                                                                              
                else:                                                                                                                       
                    raise
```

I don't know why `!d` converts to epoch time with empty string and with `!D` does not.  It is documented that way in the manpage.  That seems inconsistent.  And I think empty string and 0 mean two different things.  For example, `t_remaining` being empty string might mean the job isn't running vs 0 means there's 0 time remaining.  Similarly `t_inactive` being empty string may mean the job isn't done.

I will note that in `JobInfo` we do default:

```
    #  Default values for job properties.
    defaults = {
        "t_depend": 0.0,
        "t_run": 0.0,
        "t_cleanup": 0.0,
        "t_inactive": 0.0,
```

I think it has been this way forever, but maybe shouldn't be.  Wondering if we could default that to `""` and that means output "" vs epoch.

I will note that I'm not sure what fallout there could be with the json output option if we adjusted things.

Will ponder.

grondo replied:
> I don't know why !d converts to epoch time with empty string and with !D does not. It is documented that way in the manpage. That seems inconsistent. 

One must be a Python datetime object and the other is a string. You cant return an empty string as a datetime object.

grondo replied:
> Hmmm, I think the reason I went with !D may have been because we (I don't think) support the h hyphen output with !d. So I sort of dislike this output for the running jobs.

Well, this is a bit unsatisfying, but this allows `UtilDatetime` class to use `h`. It requires subverting `h` handling in `format_items` becuase the `UtilDatetime` formatting is done _after_ this function, then adds `h` suffix handling specifically to the `format()` method of the class.

We maybe should also have `datetime.fromtimestamp(0.)` format to an empty string which would help here:

```diff
diff --git a/src/bindings/python/flux/util.py b/src/bindings/python/flux/util.py
index befc950b4..87deb55b6 100644
--- a/src/bindings/python/flux/util.py
+++ b/src/bindings/python/flux/util.py
@@ -312,16 +312,21 @@ class UtilDatetime(datetime):
     def __format__(self, fmt):
         # The string "::" is used to split the strftime() format from
         # any Python format spec:
-        vals = fmt.split("::", 1)
+        timefmt, *spec = fmt.split("::", 1)
 
         # Call strftime() to get the formatted datetime as a string
-        result = self.strftime(vals[0])
+        result = self.strftime(timefmt)
+
+        spec = spec[0] if spec else ""
+        if spec.endswith("h"):
+            if self == datetime.fromtimestamp(0.):
+                result = "-"
+            spec = spec[:-1] + "s"
 
         # If there was a format spec, apply it here:
-        try:
-            return f"{{0:{vals[1]}}}".format(result)
-        except IndexError:
-            return result
+        if spec:
+            return f"{{0:{spec}}}".format(result)
+        return result
 
 
 def fsd(secs):
@@ -421,7 +426,7 @@ class UtilFormatter(Formatter):
             denote_truncation = True
             spec = spec[:-1]
 
-        if spec.endswith("h"):
+        if spec.endswith("h") and not isinstance(value, UtilDatetime):
             localepoch = datetime.fromtimestamp(0.0).strftime("%FT%T")
             basecases = ("", "0s", "0.0", "0:00:00", "1970-01-01T00:00:00", localepoch)
             value = "-" if str(value) in basecases else str(value)
```

grondo replied:
If this is acceptable I can propose a small PR. It would be more satisfying to handle the `h` in one place, but due to the way the formatting percolates through all the layers here (which frankly is warping my hippocampus).

chu11 replied:
Ahhh your post is very timely, I was looking into how to add `h` and fumbling and just about to message you if you had an idea on how to proceed.

`+        if spec.endswith("h") and not isinstance(value, UtilDatetime):`

This was the part I was struggling to get around, didn't think to just skip it.

PR sounds good.  Thanks.

grondo replied:
In addition to the changes to the `UtilDatetime` class to support the `h` suffix, what do you think about formatting all "zero" `UtilDatetime` objects as an empty string in the class' `format()` function @chu11?

Seems like this would be consistent with `!D` as you note, and would avoid the requirement to use `h` when you don't want to display `Dec31 ...` for unset timestamps.

chu11 replied:
> Seems like this would be consistent with !D as you note, and would avoid the requirement to use h when you don't want to display Dec31 ... for unset timestamps.

That sounds like a good idea!

chu11 commented:
re-pushed just to get initial review on fixes from comments from above + one fixup I randomly found.  Changed to WIP since this depends on PR fix for `h` converter per comment above.

- noticed that the checks for "empty things" is different between `h` and `filter_empty()`, so added a fix for that.

- added a new `compat36` lib for Python 3.6 `strsignal()`.  I ended up adding signal -> string for all signals 1 through 29.  I didn't go above that because I questioned portability of all those real time signals in the 30-40 range.  No evidence of lack of portability, just a gut feeling.  Also, oddly, one signal was only defined in Python 3.11, go figure.

chu11 commented:
now that #5169 is merged, rebased and re-pushed and removed WIP

grondo commented on src/bindings/python/flux/util.py:
```diff
@@ -426,6 +426,11 @@ def convert_field(self, value, conv):
             value = super().convert_field(value, conv)
         return value
 
+    def empty_outputs(self):
+        localepoch = datetime.fromtimestamp(0.0).strftime("%FT%T")
+        empty = ("", "0s", "0.0", "0:00:00", "1970-01-01T00:00:00", localepoch)
+        return empty
```
This function doesn't seem to use `self` so maybe it can be defined outside of this class. This will allow the unnecessary instantiation of a `UtilFormatter()` object to be dropped in `fliter_empty()` below.

grondo reviewed (commented):
Just a couple comments up for further discussion.

chu11 commented on src/bindings/python/flux/job/info.py:
```diff
@@ -541,6 +541,59 @@ def zero_remove(key):
+                return f'Exception: type={self.exception.type}{note or ""}'
+            elif self.returncode > 128:
+                signum = self.returncode - 128
+                # strsignal() is only available in Python 3.8 and up.
+                # flux-core's minimum is 3.6.  Use compat library
+                from flux.compat36 import strsignal
```
I got a vermin failure when I tried this previously, lemme give it another shot.  Perhaps the import in the middle of the file causes part of the problem.

chu11 replied:
going into vermin's code, it seems like this is just a hard require

```
    "signal.strsignal": (None, (3, 8)),
```

BUT I cannot generate this error message on a local install of vermin with 1.5.1, wondering if this could be worked around with a newer version of vermin.  playing around.

grondo replied:
well that's a bummer. Vermin makes it so you can't support multiple versions of Python? Do we really need the vermin checks? Our CI already runs all tests with multiple versions of Python.

Though maybe doing the ImportError thing is considered bad practice by Python folks? I'm having trouble forming an opinion now. Sorry for putting you through that.

chu11 commented:
rebased and re-pushed.  I went ahead and threw in some `#novermin` comments to skip these issues in vermin.  I went ahead and was "preemptive", putting `#novermin` in for something that isn't caught in vermin 1.4.2 (in our workflow) where as it is caught in version 1.5.1.  Also threw in an extra commit for a test timeout I see once in awhile.

As a complete aside, this was also caught from newer version vermin check.  So maybe could use `# novermin` too.  I'll update the vermin version in a different PR.

```
try:                                                                                                                                        
    import tomllib                                                                                                                          
except ModuleNotFoundError:                                                                                                                 
    from flux.utils import tomli as tomllib             
```

grondo reviewed (approved):
LGTM!

chu11 commented:
thanks, setting MWP
